### DISCUSSION TITLE
Ufe lights generated at runtime

### DISCUSSION THREAD

**Discussion by yolov** [Autodesk/maya-usd#2750] [General]
We've come across an issue with MayaUSD 0.19.0 and Maya 2023.2 where Ufe lights are generated at runtime in addition to the lights in a USD file. Starting a discussion to get more opinions on this, so at the point where it's no doubt considered a bug - feel free to move it to "issues".

This seems to affect MayaUSD 0.19.0 and later and only affects Maya 2023.2. Our guess is that there must be something in this particular Maya version's Ufe that allows a change in MayaUSD to do this.

We came across this when working with instanced non-geometry primitives (lights). In the above setup (0.19.0 + 2023.2) Maya would generate additional lights at runtime - we do not expect those lights to be there, but they also render wrong (at least with V-Ray), because they have noDecay enabled (which actually made it easier to spot).

It seems that `MFnAreaLight` isntances are created and are reported with names such as `ufeLightProxyShape`

Attaching some images (let me know if you can't make it out) showing different combos of MayaUSD and Maya versions compared to the original render (which renders lights before exporting them to USD).
![image (10)](https://user-images.githubusercontent.com/20951458/204571470-9c0d0180-0929-4ded-bf5e-5371c9159028.png)

Also attaching a repro file. Note that the repro file is a simplifed version of the scene you'll see in the images, so the render won't be an exact match.
The expected behavior is to get the same result when rendering:
a) using the lights from the Lights group in Maya (either delete the USD proxy or deactivate its Lights primitve group )
b) using the lights from the USD proxy (delete the Lights group in Maya or hide it)

[ufe_lights_runtime.zip](https://github.com/Autodesk/maya-usd/files/10114446/ufe_lights_runtime.zip)

**Comment by vlasovi** [comment]
Hello @yolov,
Those are the nodes we add to Maya to have light representation in the viewport. Those lights are also conveniently picked by renderers not supporting USD.
For renderers suporting USD along with Maya native data, we recommend to filter those lights with the following code (taken from MayaUSD plugin):

```
// Skip UFE nodes coming from USD runtime
// Those will be handled by USD Imaging delegate
MStatus              status;
static const MString ufeRuntimeStr = "ufeRuntime";
MPlug                ufeRuntimePlug = dagNode.findPlug(ufeRuntimeStr, false, &status);
if ((status == MS::kSuccess) && ufeRuntimePlug.asString() == "USD") {
	return;
}
```

**Reply by yolov** [reply]
Thanks, @vlasovi, that clears up why this is happening.
One more question - how does it know which renderers (don't) support USD?
Our problem is that V-Ray supports USD, but these extra lights break things for us. We're wondering what's the correct way to register V-Ray as USD-aware, f.e. do we need to add something in one of the plugInfo.json files?
Filtering the lights using the above snippet would be the backup plan.

Thanks in advance.

Alex

**Comment by vlasovi** [comment]
@yolov To my knowledge, there doesn't exist any flag like you mentioned.

**Comment by sshklifov** [comment]
Hi everyone, @yolov and @vlasovi.

There probably doesn't exist such a flag, yes. But I think it should. While adding the linked code and filtering out the ufe lights is trivial, this will, in my mind, lead to bugs in the future. For example, we might decide to create a command that exports only lights to USD, test with a MayaUSD version that doesn't have UFE and realize too late that this snippet has to be added. Or, another example, say in the future another UFE node is created that isn't a light, the snippet will be missing in that particular code segment and we run into the same problem.

Is it not possible to add a setting inside plugInfo.json that controlls UFE the same way that there is a setting that controls mayaUSDExport command?

I think that there are potential issues as it stands now. I would like to hear what you think, as I am not familiar with UFE or whether the solution I proposed is a good one.

Thanks,
Stefan
